### PR TITLE
Binary Image Segmentation using MaskRCNN with Resizing

### PR DESCRIPTION
**Pull request by DamnikJain1**
# Description

Image segmentation using pre-trained MaskRCNN, then using the mask, applied binary segmentation to mask out the child from the background. The final result is in image format with child as 1 and background as 0 which I am display on the system.

List any dependencies that are required for this change.
`pip install torch torchvision`

User story: https://dev.azure.com/cgmorg/ChildGrowthMonitor/_workitems/edit/301

# How Has This Been Tested?

Main file: `maskrcnn_resizeimage.ipynb`
I have tested this out on the anon_rgb_training dataset on Azure on different poses. Poses like 100 & 200 were giving good results, where the child is aligned straight to the camera. In some of the cases, the mask was not able to cover the child fully, some portion was left out. You can have a look at the results by opening `src/common/reliability/Image-Segmentation/maskrcnn_resizeimage.ipynb`

# Checklist:

- [x] I have tested the code on Azure under Users/jaindamnik203 folder
- [x] The code is formatted & aligned.
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests
- [x] New and existing unit tests pass locally with my changes
- [ ] My Code is review by 1 People

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -17,3 +17,5 @@ pytest==5.4.2
 pytest-cov==2.10.0
 pytest-azurepipelines==0.8.0
 xmltodict==0.12.0
+torch
```
Comment (markus-hinsche) [inline requirements.txt:20]: Please add the versions you use, ideally the most up-to-date versions

File: src/common/reliability/Image-Segmentation/imgseg/predict.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
+import logging
+
+import numpy
+import numpy as np
+from PIL import Image, ImageFont
+# from PIL.ImageDraw import Draw
```
Comment (markus-hinsche) [inline src/common/reliability/Image-Segmentation/imgseg/predict.py:6]: Please remove all commented code in this pull request

File: src/common/reliability/Image-Segmentation/maskrcnn_resizeimage.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+# coding: utf-8
+import sys
+import numpy as np
+from imgseg.predict import predict, show
+from torchvision.models.detection import maskrcnn_resnet50_fpn
+from PIL import Image
+import time
+# from IPython.display import display
+# from utils.ipstartup import *
+sys.path.insert(0, "utils")
```
Comment (markus-hinsche) [inline src/common/reliability/Image-Segmentation/maskrcnn_resizeimage.py:11]: Is this line needed?

File: temp.txt [hunk 4]
```diff
@@ -0,0 +1,476 @@
+absl-py==0.10.0
```
Comment (markus-hinsche) [inline temp.txt:1]: Do we need this file? Otherwise please remove it

File: src/common/reliability/Image-Segmentation/maskrcnn_resizeimage.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        newsize = (int(image.size[0] / factor), int(image.size[1] / factor))
+        rimage = image.resize(newsize)
+    start_time = time.time()
+    print("After Resizing, Image Dimension: ", rimage.size)
+    out = predict(rimage, model)
+    print("Time: %s s" % (time.time() - start_time))
+    return (rimage, out)
+
+#File path
+def processImage(filepath):
```
Comment (markus-hinsche) [inline src/common/reliability/Image-Segmentation/maskrcnn_resizeimage.py:29]: Please make all own function names in this PR snake_case instead of camelCase

```suggestion
def process_image(filepath):
```

File: src/common/reliability/Image-Segmentation/utils/azurestartup.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
+import shlex
```
Comment (markus-hinsche) [inline src/common/reliability/Image-Segmentation/utils/azurestartup.py:1]: Is this file used somewhere? Otherwise please remove it

File: src/common/reliability/Image-Segmentation/utils/cellevents.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
+""" jupyter extension:
```
Comment (markus-hinsche) [inline src/common/reliability/Image-Segmentation/utils/cellevents.py:1]: The modules `cellevents.py`,  `ipstartup.py`, `defaultlog.py` see to be unrelated to the purpose of this PR which is image segmentation. Can we remove them?

File: src/common/reliability/Image-Segmentation/tests/test_maskrcnn_resizeimage.py [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+sys.path.append(str(Path(__file__).parents[1]))
+
+from maskrcnn_resizeimage import processImage  # noqa: E402
+
+IMAGE_FNAME = "rgb_test.jpg"
+
+def test_maskrcnn_resizeimage():
+    source_path = str(Path(__file__).parent / IMAGE_FNAME)
+    res = processImage(source_path)
+    assert res is not None
```
Comment (markus-hinsche) [inline src/common/reliability/Image-Segmentation/tests/test_maskrcnn_resizeimage.py:15]: This check is not sufficient. This only checks that something happens. 
Actually it would be great to check if the correct/expected thing happen. 

Example in pseudocode: 
```
def test_maskrcnn_resizeimage():
    source_path = str(Path(__file__).parent / IMAGE_FNAME)
    segmented_image = segment_image(source_path)
    
    # assert the corner upper pixel belongs to the background
    assert segmented_image[0,0] = 0

    # assert the middle upper pixel belongs to the foreground
    w, h = segmented_image.size
    assert segmented_image[int(w/2), int(h/2)] == 1
```

File: src/common/reliability/Image-Segmentation/maskrcnn_resizeimage.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+from PIL import Image
+import time
+# from IPython.display import display
+# from utils.ipstartup import *
+sys.path.insert(0, "utils")
+
+model = maskrcnn_resnet50_fpn(pretrained=True)
+
+# Applies MaskRCNN on downscaled image, by default the factor is 10x
+def predictByResize(image, flag=0, factor=10):
```
Comment (markus-hinsche) [inline src/common/reliability/Image-Segmentation/maskrcnn_resizeimage.py:16]: Use docstrings to document functions
```suggestion
def predictByResize(image, flag=0, factor=10):
    """Applies MaskRCNN on downscaled image, by default the factor is 10x"""
```

**Review by markus-hinsche — CHANGES_REQUESTED** [review]
Thanks Damnik for your contribution. This is great!

I commented a few more places for improvement of the code, most noticeably:
- we can remove all code we don't need (comments, unused modules, etc.)
- we can add unit test that really check if the functionality works as expected

Damnik: It would be great if you could go through the comments and see if you can still address some of the issues. 

After this I will see how I can merge this. 
Great thanks again!

**Comment by DamnikJain1** [comment]
Hi @markus-hinsche , I have gone through your comments, done the necessary changes and committed the code.

**Comment by markus-hinsche** [comment]
I will continue this work in this PR: https://github.com/Welthungerhilfe/cgm-ml/pull/173

**Comment by markus-hinsche** [comment]
@DamnikJain1 Thanks a lot for your contribution, we will merge it shortly
